pkg/hive/cell/structured.go

```diff
@@ -0,0 +1,384 @@
+func TestScope(hr HealthReporter) Scope {
+	switch hr.(type) {
+	case *reporter, *subReporter:
+		panic("test scope should only be used on test health reporter implementations")
+	}
+	return rootScope(hr)
+}
+
+// TODO: Add flush function.
+func rootScope(hr HealthReporter) Scope {
+	r := &subReporter{
+		base: &subreporterBase{
+			hr:           hr,
+			idToChildren: map[string]children{},
+			nodes:        map[string]*node{},
+		},
+	}
+	// create root node, required in case reporters are created without any subscopes.
+	r.id = r.base.addChild("", "", false)
+	r.base.rootID = r.id
+
+	realize := func() {
+		r.base.Lock()
+		defer r.base.Unlock()
+		statusTree := r.base.getStatusTreeLocked(r.base.rootID)
+		if statusTree.allOk() {
+			r.base.hr.OK(statusTree.toString(2))
+		} else {
+			r.base.hr.Degraded(statusTree.toString(2), nil)
+		}
+	}
+	t, err := trigger.NewTrigger(trigger.Parameters{
+		MinInterval: time.Millisecond * 100,
+		TriggerFunc: func(reasons []string) {
+			realize()
+		},
+	})
+	r.realize = t.Trigger
+	if err != nil {
+		panic("BUG: could not create reporter root scope trigger")
+	}
+	return &scope{subReporter: r}
+}
+
+type Scope interface {
+	Name() string
+	Close()
+	scope() *subReporter
+}
+
+type scope struct {
+	*subReporter
+}
+
+func (s *scope) Close() {
+	s.base.Lock()
+	s.base.removeTreeLocked(s.id)
+	s.base.Unlock()
+	s.realize()
+}
+
+func (s *scope) scope() *subReporter {
+	return s.subReporter
+}
+
+func (s *scope) Name() string {
+	return s.name
+}
+
+func createSubScope(parent Scope, name string) *scope {
+	return &scope{
+		subReporter: getSubReporterV2(parent, name, false),
+	}
+}
+
+func getSubReporterV2(parent Scope, name string, isReporter bool) *subReporter {
+	return scopeFromParent(parent, name, isReporter)
+}
+
+func scopeFromParent(parent Scope, name string, isReporter bool) *subReporter {
+	r := parent.scope()
+	r.base.Lock()
+	defer r.base.Unlock()
+
+	// If such a reporter already exists at this scope, we just return the same reporter
+	// by recreating the subreporter.
+	for cid := range r.base.idToChildren[r.id] {
+		child := r.base.nodes[cid]
+		if child.name == name {
+			return &subReporter{
+				base:    r.base,
+				id:      cid,
+				realize: r.realize,
+				name:    name,
+			}
+		}
+	}
+
+	id := r.base.addChild(r.id, name, isReporter)
+
+	return &subReporter{
+		base:    r.base,
+		id:      id,
+		realize: r.realize,
+		name:    name,
+	}
+}
+
+type subreporterBase struct {
+	lock.Mutex
+	hr           HealthReporter
+	idToChildren map[string]children
+	nodes        map[string]*node
+	rootID       string
+}
+
+func (s *subreporterBase) addNode(n *node) {
+	if _, ok := s.idToChildren[n.parentID]; !ok {
+		s.idToChildren[n.parentID] = children{}
+	}
+	s.idToChildren[n.parentID][n.id] = struct{}{}
+	s.idToChildren[n.id] = children{}
+	s.nodes[n.id] = n
+}
+
+func (s *subreporterBase) addChild(pid string, name string, isReporter bool) string {
+	id := name + "_" + uuid.New().String()
```

Comment:
Wonder if a global counter would be nicer here. It would also inform in what order these were added which might be useful. And looks like we're sorting these later, so using a global counter would probably preserve the order across restarts. Do these ids need to be unique outside this process?

Reply:
> Do these ids need to be unique outside this process?

For now, no.

Makes sense, I've changed this to just use a counter within the reporter. I've changed it to use a counter within the reporter base scope and prefix the ID with the uid, that should preserve order when realizing since we sort by IDs.

What would be the benefit of having this be global across all reporters?